File: C#/Crawler.cs
using System.Text.RegularExpressions;

namespace SecurityTester;

/// <summary>
/// Represents a single discovered page during crawling.
/// </summary>
public class PageInfo
{
    public string   Url         { get; set; } = "";
    public int      Status      { get; set; }
    public int      Depth       { get; set; }
    public string   ContentType { get; set; } = "";
    public List<string> QueryParams { get; set; } = new();
    public bool     HasForms    { get; set; }
}

/// <summary>
/// Represents a discovered HTML form.
/// </summary>
public class FormInfo
{
    public string       Action  { get; set; } = "";
    public string       Method  { get; set; } = "GET";
    public List<string> Fields  { get; set; } = new();
    public string       FoundOn { get; set; } = "";
}

/// <summary>
/// Full site map produced by the crawler.
/// </summary>
public class SiteMap
{
    public string           StartUrl    { get; set; } = "";
    public List<PageInfo>   Pages       { get; set; } = new();
    public List<FormInfo>   Forms       { get; set; } = new();
    public List<string>     Endpoints   { get; set; } = new();  // URLs without query params
    public List<PageInfo>   ParamUrls   { get; set; } = new();  // URLs that have query params
    public int              TotalFound  => Pages.Count;
}

/// <summary>
/// Web crawler — visits the root URL and follows all internal links recursively.
/// Extracts pages, forms, endpoints, and query parameters for security testing.
/// </summary>
public class Crawler
{
    private static readonly string[] IgnoredExtensions =
    {
        ".png", ".jpg", ".jpeg", ".gif", ".svg", ".ico", ".webp",
        ".pdf", ".zip", ".tar", ".gz", ".mp4", ".mp3", ".woff",
        ".woff2", ".ttf", ".eot", ".css", ".map", ".js"
    };

    private readonly HttpClient _client;
    private readonly int        _maxPages;
    private readonly int        _maxDepth;
    private readonly int        _delayMs;
    private readonly bool       _verbose;

    public Crawler(HttpClient client, int maxPages = 100, int maxDepth = 4, int delayMs = 100, bool verbose = true)
    {
        _client   = client;
        _maxPages = maxPages;
        _maxDepth = maxDepth;
        _delayMs  = delayMs;
        _verbose  = verbose;
    }

    public async Task<SiteMap> CrawlAsync(string startUrl)
    {
        var siteMap  = new SiteMap { StartUrl = startUrl };
        var visited  = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var queue    = new Queue<(string Url, int Depth)>();
        var endpoints = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        queue.Enqueue((startUrl, 0));

        if (_verbose)
        {
            Console.WriteLine($"\n   🕷️  Crawler starting...");
            Console.WriteLine($"   Max pages : {_maxPages}");
            Console.WriteLine($"   Max depth : {_maxDepth}\n");
        }

        while (queue.Count > 0 && visited.Count < _maxPages)
        {
            var (url, depth) = queue.Dequeue();

            if (visited.Contains(url))  continue;
            if (depth > _maxDepth)      continue;

            visited.Add(url);

            if (_verbose)
                Console.Write($"\r   Crawling [{visited.Count}/{_maxPages}]: {url.PadRight(80).Substring(0, Math.Min(url.Length, 75))}...");

            string? body        = null;
            int     status      = 0;
            string  contentType = "";

            try
            {
                var res = await _client.GetAsync(url);
                status      = (int)res.StatusCode;
                contentType = res.Content.Headers.ContentType?.MediaType ?? "";

                if (contentType.Contains("text/html") || contentType.Contains("application/json"))
                    body = await res.Content.ReadAsStringAsync();
            }
            catch { continue; }

            if (status == 404) continue;

            // Build page record
            var pageInfo = new PageInfo
            {
                Url         = url,
                Status      = status,
                Depth       = depth,
                ContentType = contentType,
                QueryParams = ExtractQueryParams(url),
            };

            // Track endpoint (path without query string)
            if (Uri.TryCreate(url, UriKind.Absolute, out var parsedUri))
            {
                var ep = $"{parsedUri.Scheme}://{parsedUri.Host}{parsedUri.AbsolutePath}";
                endpoints.Add(ep.TrimEnd('/'));
            }

            // Track pages with query params
            if (pageInfo.QueryParams.Count > 0)
                siteMap.ParamUrls.Add(pageInfo);

            // Parse HTML
            if (body != null && contentType.Contains("text/html"))
            {
                // Extract forms
                var forms = ExtractForms(url, body);
                if (forms.Count > 0)
                {
                    pageInfo.HasForms = true;
                    siteMap.Forms.AddRange(forms);
                }

                // Extract and queue links
                if (depth < _maxDepth)
                {
                    var links = ExtractLinks(url, body);
                    foreach (var link in links)
                    {
                        if (!visited.Contains(link))
                            queue.Enqueue((link, depth + 1));
                    }
                }
            }

            siteMap.Pages.Add(pageInfo);

            if (_delayMs > 0 && queue.Count > 0)
                await Task.Delay(_delayMs);
        }

        siteMap.Endpoints = endpoints.ToList();

        if (_verbose)
        {
            Console.WriteLine();
            Console.WriteLine($"\n   ✅ Crawl complete: {siteMap.Pages.Count} pages, {siteMap.Forms.Count} forms, {siteMap.Endpoints.Count} endpoints\n");
        }

        return siteMap;
    }

    // ── Private helpers ───────────────────────────────────────

    private List<string> ExtractLinks(string baseUrl, string html)
    {
        var links   = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var baseUri = new Uri(baseUrl);

        // href, action, src attributes
        var attrPattern = new Regex(@"(?:href|action|src)=[""']([^""'#>]+)[""']", RegexOptions.IgnoreCase);
        foreach (Match m in attrPattern.Matches(html))
            links.Add(m.Groups[1].Value.Trim());

        // API endpoints in JS (fetch, axios calls)
        var apiPattern = new Regex(@"(?:fetch|get|post|put|delete|axios)\s*\(\s*[""'`](/[^""'`\s]+)[""'`]", RegexOptions.IgnoreCase);
        foreach (Match m in apiPattern.Matches(html))
            links.Add(m.Groups[1].Value.Trim());

        // Resolve to absolute, filter to same origin
        var resolved = new List<string>();
        foreach (var link in links)
        {
            try
            {
                var abs = new Uri(baseUri, link);
                if (abs.Host != baseUri.Host) continue;

                // Skip ignored extensions
                var path = abs.AbsolutePath.ToLower();
                if (IgnoredExtensions.Any(ext => path.EndsWith(ext))) continue;

                // Remove fragment
                var clean = abs.GetLeftPart(UriPartial.Query);
                resolved.Add(clean);
            }
            catch { }
        }

        return resolved;
    }

    private List<FormInfo> ExtractForms(string pageUrl, string html)
    {
        var forms       = new List<FormInfo>();
        var formPattern = new Regex(@"<form[^>]*>([\s\S]*?)<\/form>", RegexOptions.IgnoreCase);
        var baseUri     = new Uri(pageUrl);

        foreach (Match formMatch in formPattern.Matches(html))
        {
            var formHtml = formMatch.Value;

            var actionMatch = Regex.Match(formHtml, @"action=[""']([^""'>]+)[""']", RegexOptions.IgnoreCase);
            var methodMatch = Regex.Match(formHtml, @"method=[""']([^""'>]+)[""']", RegexOptions.IgnoreCase);

            string action;
            try   { action = new Uri(baseUri, actionMatch.Success ? actionMatch.Groups[1].Value : "").ToString(); }
            catch { action = pageUrl; }

            var method = methodMatch.Success ? methodMatch.Groups[1].Value.ToUpper() : "GET";
            var fields = new List<string>();

            foreach (Match input in Regex.Matches(formHtml, @"<input[^>]*name=[""']([^""'>]+)[""']", RegexOptions.IgnoreCase))
                fields.Add(input.Groups[1].Value);
            foreach (Match sel in Regex.Matches(formHtml, @"<select[^>]*name=[""']([^""'>]+)[""']", RegexOptions.IgnoreCase))
                fields.Add(sel.Groups[1].Value);
            foreach (Match ta in Regex.Matches(formHtml, @"<textarea[^>]*name=[""']([^""'>]+)[""']", RegexOptions.IgnoreCase))
                fields.Add(ta.Groups[1].Value);

            forms.Add(new FormInfo { Action = action, Method = method, Fields = fields, FoundOn = pageUrl });
        }

        return forms;
    }

    private static List<string> ExtractQueryParams(string url)
    {
        try
        {
            var uri    = new Uri(url);
            var query  = System.Web.HttpUtility.ParseQueryString(uri.Query);
            return query.AllKeys.Where(k => k != null).Select(k => k!).ToList();
        }
        catch { return new List<string>(); }
    }
}

File: js/crawler.js
const https = require("https");
const http  = require("http");
const { URL } = require("url");

/**
 * Web Crawler
 * - Visits root URL and all internal links recursively
 * - Respects depth limit and max pages to avoid infinite loops
 * - Extracts: page URLs, form actions, API endpoints, query params
 * - Returns structured site map for security tests to consume
 */

const IGNORED_EXTENSIONS = [
  ".png", ".jpg", ".jpeg", ".gif", ".svg", ".ico", ".webp",
  ".pdf", ".zip", ".tar", ".gz", ".mp4", ".mp3", ".woff",
  ".woff2", ".ttf", ".eot", ".css", ".map"
];

const MAX_PAGES  = 100;  // max pages to crawl
const MAX_DEPTH  = 4;    // max link depth from root
const TIMEOUT    = 6000;
const DELAY_MS   = 100;  // polite delay between requests

async function fetchPage(url) {
  return new Promise((resolve) => {
    try {
      const parsed = new URL(url);
      const lib    = parsed.protocol === "https:" ? https : http;
      const req    = lib.get(url, {
        timeout: TIMEOUT,
        rejectUnauthorized: false,
        headers: { "User-Agent": "SecurityTester-Crawler/1.0 (Educational)" }
      }, (res) => {
        // Only parse HTML pages
        const ct = res.headers["content-type"] || "";
        if (!ct.includes("text/html") && !ct.includes("application/json")) {
          res.resume();
          resolve({ status: res.statusCode, body: "", headers: res.headers, contentType: ct });
          return;
        }
        let body = "";
        res.on("data", (c) => {
          body += c;
          if (body.length > 500000) { req.destroy(); } // cap at 500KB
        });
        res.on("end", () => resolve({ status: res.statusCode, body, headers: res.headers, contentType: ct }));
      });
      req.on("error",   () => resolve(null));
      req.on("timeout", () => { req.destroy(); resolve(null); });
    } catch { resolve(null); }
  });
}

function extractLinks(baseUrl, html) {
  const links = new Set();
  const base  = new URL(baseUrl);

  // href links
  const hrefMatches = html.matchAll(/href=["']([^"'#>]+)["']/gi);
  for (const m of hrefMatches) links.add(m[1].trim());

  // form actions
  const actionMatches = html.matchAll(/action=["']([^"'>]+)["']/gi);
  for (const m of actionMatches) links.add(m[1].trim());

  // src attributes (scripts, iframes)
  const srcMatches = html.matchAll(/src=["']([^"'>]+)["']/gi);
  for (const m of srcMatches) links.add(m[1].trim());

  // API endpoints in JS (fetch, axios, XMLHttpRequest)
  const apiMatches = html.matchAll(/(?:fetch|get|post|put|delete|axios)\s*\(\s*["'`]([/][^"'`\s]+)["'`]/gi);
  for (const m of apiMatches) links.add(m[1].trim());

  // Resolve to absolute URLs, filter to same origin
  const resolved = new Set();
  for (const link of links) {
    try {
      const abs = new URL(link, baseUrl);
      // Same host only
      if (abs.hostname !== base.hostname) continue;
      // Skip ignored extensions
      const path = abs.pathname.toLowerCase();
      if (IGNORED_EXTENSIONS.some(ext => path.endsWith(ext))) continue;
      // Clean URL (remove fragment)
      abs.hash = "";
      resolved.add(abs.toString());
    } catch { }
  }

  return resolved;
}

function extractForms(url, html) {
  const forms = [];
  const formMatches = html.matchAll(/<form[^>]*>([\s\S]*?)<\/form>/gi);

  for (const formMatch of formMatches) {
    const formHtml = formMatch[0];

    // Get action
    const actionMatch = formHtml.match(/action=["']([^"'>]+)["']/i);
    const methodMatch = formHtml.match(/method=["']([^"'>]+)["']/i);

    const action = actionMatch
      ? new URL(actionMatch[1], url).toString()
      : url;
    const method = methodMatch ? methodMatch[1].toUpperCase() : "GET";

    // Get input fields
    const fields = [];
    const inputMatches = formHtml.matchAll(/<input[^>]*name=["']([^"'>]+)["'][^>]*>/gi);
    for (const m of inputMatches) fields.push(m[1]);

    const selectMatches = formHtml.matchAll(/<select[^>]*name=["']([^"'>]+)["'][^>]*>/gi);
    for (const m of selectMatches) fields.push(m[1]);

    const textareaMatches = formHtml.matchAll(/<textarea[^>]*name=["']([^"'>]+)["'][^>]*>/gi);
    for (const m of textareaMatches) fields.push(m[1]);

    forms.push({ action, method, fields, foundOn: url });
  }

  return forms;
}

function extractQueryParams(url) {
  try {
    const parsed = new URL(url);
    return [...parsed.searchParams.keys()];
  } catch { return []; }
}

async function crawl(startUrl, options = {}) {
  const {
    maxPages = MAX_PAGES,
    maxDepth = MAX_DEPTH,
    verbose  = true,
  } = options;

  const visited    = new Set();
  const queue      = [{ url: startUrl, depth: 0 }];
  const pages      = [];   // all discovered pages
  const forms      = [];   // all discovered forms
  const endpoints  = new Set(); // unique endpoints (URL without params)
  const paramUrls  = [];   // URLs that have query parameters

  if (verbose) {
    console.log(`\n   🕷️  Crawler starting...`);
    console.log(`   Max pages : ${maxPages}`);
    console.log(`   Max depth : ${maxDepth}\n`);
  }

  while (queue.length > 0 && visited.size < maxPages) {
    const { url, depth } = queue.shift();

    if (visited.has(url))   continue;
    if (depth > maxDepth)   continue;

    visited.add(url);

    if (verbose) process.stdout.write(`   Crawling [${visited.size}/${maxPages}]: ${url.slice(0, 70)}...\r`);

    const res = await fetchPage(url);
    if (!res || res.status === 404) continue;

    // Record this page
    const pageData = {
      url,
      status:      res.status,
      depth,
      contentType: res.contentType,
      queryParams: extractQueryParams(url),
      hasForms:    false,
    };

    // Track endpoint (path without query params)
    try {
      const parsed = new URL(url);
      parsed.search = "";
      endpoints.add(parsed.toString());
    } catch { }

    // Track URLs with query params for injection testing
    if (pageData.queryParams.length > 0) {
      paramUrls.push({ url, params: pageData.queryParams });
    }

    // Extract forms
    if (res.body && res.contentType.includes("text/html")) {
      const pageForms = extractForms(url, res.body);
      if (pageForms.length > 0) {
        pageData.hasForms = true;
        forms.push(...pageForms);
      }

      // Extract and queue new links
      if (depth < maxDepth) {
        const links = extractLinks(url, res.body);
        for (const link of links) {
          if (!visited.has(link)) {
            queue.push({ url: link, depth: depth + 1 });
          }
        }
      }
    }

    pages.push(pageData);

    if (DELAY_MS > 0 && queue.length > 0) {
      await new Promise(r => setTimeout(r, DELAY_MS));
    }
  }

  if (verbose) {
    console.log(`\n   ✅ Crawl complete: ${pages.length} pages, ${forms.length} forms, ${endpoints.size} endpoints\n`);
  }

  return {
    startUrl,
    pages,
    forms,
    endpoints:  [...endpoints],
    paramUrls,
    totalFound: pages.length,
  };
}

module.exports = { crawl };
